mods/aerexplorer.wh.cpp

```diff
@@ -0,0 +1,3068 @@
+    {
+        settings.npst = NPST_VISTA;
+    }
+    else if (0 == wcscmp(szNpst, L"seven"))
+    {
+        settings.npst = NPST_SEVEN;
+    }
+    else
+    {
+        settings.npst = NPST_DEFAULT;
+    }
+    Wh_FreeStringSetting(szNpst);
+
+    LPCWSTR szTbst = Wh_GetStringSetting(L"tbst");
+    if (0 == wcscmp(szTbst, L"seven"))
+    {
+        settings.tbst = TBST_SEVEN;
+    }
+    else if (0 == wcscmp(szTbst, L"ten"))
+    {
+        settings.tbst = TBST_TEN;
+    }
+    else if (0 == wcscmp(szTbst, L"tennew"))
+    {
+        settings.tbst = TBST_TENNEW;
+    }
+    else if (0 == wcscmp(szTbst, L"custom"))
+    {
+        settings.tbst = TBST_CUSTOM;
+    }
+    else
+    {
+        settings.tbst = TBST_DEFAULT;
+    }
+    Wh_FreeStringSetting(szTbst);
+}
+
+#ifdef _WIN64
+#   define PATHCACHE_VALNAME L"last-comctl32-v6-path"
+#else
+#   define PATHCACHE_VALNAME L"last-comctl32-v6-path-wow64"
+#endif
+
+#define COMCTL_582_SEARCH    L"microsoft.windows.common-controls_6595b64144ccf1df_5.82"
+
+/* Load the ComCtl32 module */
+HMODULE LoadComCtlModule(void)
+{
+    HMODULE hComCtl = LoadLibraryW(L"comctl32.dll");
+    if (!hComCtl)
+    {
+        return NULL;
+    }
+
+    WCHAR szPath[MAX_PATH];
+    GetModuleFileNameW(hComCtl, szPath, MAX_PATH);
+
+    WCHAR szv6Path[MAX_PATH];
+    BOOL bNoCache = FALSE;
+    if (!Wh_GetStringValue(PATHCACHE_VALNAME, szv6Path, MAX_PATH))
+    {
+        bNoCache = TRUE;
+    }
+
+    /**
+      * the !bNoCache check here is nested because we only want to fall through
+      * to the cacher if the current comctl32 path is NOT 5.82.
+      */
+    if (wcsstr(szPath, COMCTL_582_SEARCH)
+    || wcsstr(szPath, L"\\Windows\\System32")
+    || wcsstr(szPath, L"\\Windows\\SysWOW64"))
+    {
+        if (!bNoCache)
+        {
+            hComCtl = LoadLibraryW(szv6Path);
+        }
+    }
+    else if (bNoCache || wcsicmp(szPath, szv6Path))
+    {
+        Wh_SetStringValue(PATHCACHE_VALNAME, szPath);
+    }
+
+    return hComCtl;
+}
+
+struct CMWF_SYMBOL_HOOK {
+    std::vector<std::wstring> symbols;
+    void** pOriginalFunction;
+    void* hookFunction = nullptr;
+    bool optional = false;
+    void **pSharedMemoryCache = nullptr;
+
+    template <typename Prototype>
+    CMWF_SYMBOL_HOOK(
+            std::vector<std::wstring> symbols,
+            Prototype **originalFunction,
+            std::type_identity_t<Prototype *> hookFunction = nullptr,
+            bool optional = false,
+            void *pSharedMemoryCache = nullptr
+    ) : symbols(std::move(symbols)),
+        pOriginalFunction((void **)originalFunction),
+        hookFunction((void *)hookFunction),
+        optional(optional),
+        pSharedMemoryCache((void **)pSharedMemoryCache) {}
+
+    CMWF_SYMBOL_HOOK() = default;
+};
+
+/*
+ * CmwfHookSymbols: A custom hook wrapper which allows storing symbol hook results in the
+ *                  module's shared memory for faster access.
+ */
+bool CmwfHookSymbols(
```

Comment:
I don't mind, but I hardly believe that it has any effect on performance. Did you measure it?

Reply:
> I don't mind, but I hardly believe that it has any effect on performance. Did you measure it?

It's been a while since this was added to Aerexplorer, but if I remember correctly, initialization with this is noticeably faster than without, considering the large amount of symbol hooks this mod uses.

Reply:
Perhaps it was compared to loading symbols each time, using `Wh_FindFirstSymbol`/`Wh_FindNextSymbol` and not `WindhawkUtils::HookSymbols` which has a caching mechanism.

Reply:
Aerexplorer has used `WindhawkUtils` from the start.

Reply:
Versions before 1.4 (1.3, 1.3.1) also had a bug with caching of 32-bit and 64-bit processes overriding each other. Maybe that's why you got slowness. If not, I'm not sure what it was, but if you can figure it out, let me know, I'd rather improve it in Windhawk for all mods.

BTW there's the issue of all processes loading the symbols simultaneously which is not ideal as well, and I have it on my list to improve it, but it happens only at the first time before symbols are loaded, and this shared memory trick doesn't improve it.

Reply:
I was looking at symbol usage of various mods, and stumbled upon the following in Aerexplorer:
```cpp
    /* This needs to be hooked EARLY. */
    const CMWF_SYMBOL_HOOK isCplHooks[] = {/*...*/};
        
    if (!CmwfHookSymbols(
        hExplorerFrame,
        isCplHooks,
        ARRAYSIZE(isCplHooks)
    ))
    {/*...*/}

    const CMWF_SYMBOL_HOOK efHooks[] = {/*...*/};

    if (!CmwfHookSymbols(
        hExplorerFrame,
        efHooks,
        ARRAYSIZE(efHooks)
    ))
    {/*...*/}
```

Running `WindhawkUtils::HookSymbols` more than once for the same module isn't a good idea, since the cache will be overridden every time, and so each call will enumerate all symbols, which can be slow. Perhaps that's why you saw that using the `CmwfHookSymbols` wrapper helps in your case. I suggest to merge the two hook arrays for ExplorerFrame.dll and I think it's also better to avoid the `CmwfHookSymbols` wrapper since I don't believe it has any positive effect on performance.

Regarding your "This needs to be hooked EARLY" comment, your separation has no effect, as the hooks are only applied once `Wh_ModInit` returns. See [the Mod lifetime wiki](https://github.com/ramensoftware/windhawk/wiki/Mod-lifetime) for details.